Introduce create ticket script

oshoval opened the pull request:
The script automates github issue to Jita ticket mirroring.

oshoval committed:
Introduce create ticket script

The script automates github issue to Jita ticket mirroring.

Signed-off-by: Or Shoval <oshoval@redhat.com>

EdDev commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+FROM fedora
```
I suggest you statically fix the fedora version.

EdDev commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+    pip
+
+RUN pip install requests
+RUN pip install jira
```
It is better to do this in one "run", as each creates a new storage layer.
At the end it is also common to cleanup the cache: `dnf clean all`. You may also want to do `rm -rf /var/cache/yum`, I am unsure if it is needed.

EdDev commented on LICENSE:
```diff
@@ -186,7 +186,7 @@
       identification within third-party archives.
 
-   Copyright [yyyy] [name of copyright owner]
+   Copyright [2021] [The KubeVirt Authors]
```
KubeVirt?

oshoval replied:
Hmm ye, it was requested for CNV Network project,
what do you suggest if crucial ?

EdDev replied:
This is not under the kubevirt org, so you can copyright to Red Hat Inc. I guess.

EdDev commented on README.md:
```diff
@@ -1,2 +1,43 @@
 # github2jira
 Scrap github issues and create Jira tickets
+
```
A first time visitor will most likely be glad to understand what this project does and why it is useful. Consider adding some info about it to market/advertise it.

EdDev commented on README.md:
````diff
@@ -1,2 +1,43 @@
+export GITHUB_TOKEN=<..> # see "One time configuration" section
+```
+
+2. Run `./create_ticket.py` in order to fetch github issues and create a ticket for them
````
Is there a dry-run option? This may be useful for debugging and checking what will actually be done before really doing it.

oshoval replied:
yes there is 
it doesn't create tickets on Jira, but does all beside that
https://github.com/oshoval/github2jira/pull/1/files#diff-63469d0fb919a44c962153d5288c638b1e9e5184a30d002a58eeb8a269e9878aR78

https://github.com/oshoval/github2jira/pull/1/files#diff-63469d0fb919a44c962153d5288c638b1e9e5184a30d002a58eeb8a269e9878aR32

EdDev replied:
Great. Please document it.

EdDev commented on README.md:
```diff
@@ -1,2 +1,43 @@
+
+2. Run `./create_ticket.py` in order to fetch github issues and create a ticket for them
+
+## Running as k8s payload
```
Can you add some text about what this actually means? Before doing anything, one may want to know what is it for.

EdDev commented on common.py:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python3
```
- Please add license header.
- You better call this file `main`.
- I do not really understand why it is needed at all. It does nothing.

EdDev commented on common.py:
```diff
@@ -0,0 +1,21 @@
+
+
+def main():
+    print("common self test")
```
Leftover?

EdDev commented on common.py:
```diff
@@ -0,0 +1,21 @@
+    if value is None:
+        print(f"Error: cant find {name}, exiting")
+        sys.exit(1)
+    return value
```
I do not understand why this function is useful, it does nothing special.

It is also not nice to exit in the middle like this, the exit points need to be clear and not a surprise.

EdDev commented on create_ticket.py:
```diff
@@ -0,0 +1,91 @@
+import sys
+import argparse
+
+from common import get_envvar
```
You should not import from `main`, the dependency should be in reverse.

Also, it is common practice to arrange the imports in these groups (and order):
- std
- 3rd party
- self/project

oshoval replied:
note that `create_ticket.py` is the main file, the common is a lib (that i will consider dropping according the comments)

EdDev replied:
`main.go` should have almost nothing except calling the lib.

EdDev commented on create_ticket.py:
```diff
@@ -0,0 +1,91 @@
+# process upto x weeks back
+max_delta_weeks = 4
+# how many tickets can be opened on each cycle
+flood_protection_limit = 3
```
If these are constants, use capital letters to declare that they should not be modified.

EdDev commented on create_ticket.py:
```diff
@@ -0,0 +1,91 @@
+
+
+if __name__ == "__main__":
+    main()
```
You should have only a single `main` in your project, and it usually should not be mixed with the logic.

EdDev commented on create_ticket.py:
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
```
The module name should be a noun, not a verb.
I guess this will then fit `ticket`.

EdDev commented on create_ticket.py:
```diff
@@ -0,0 +1,91 @@
+    github_token = get_envvar("GITHUB_TOKEN")
+    github_owner = get_envvar("GITHUB_OWNER")
+    github_repo = get_envvar("GITHUB_REPO")
+    github_label = get_envvar("GITHUB_LABEL")
```
I suggest you create a `config` module, where you:
- Create a class of `Config`.
- Define the env-var names as "const" at the class level.
- Define instance variables.
- Define a `load` method that places the values in the instance variables.
- Read these values directly from the attributes.

it should look like this (more or less):
```
config = config.Config()
config.Load()

foo(config.server, config.username, ....)
```

EdDev commented on create_ticket.py:
```diff
@@ -0,0 +1,91 @@
+    return False
+
+
+def loop(jira, github, dryrun):
```
- Please define functions so callees are below callers, it helps with reading the code better (from abstract to details).
- I do not think this name is valid, you need to find a better one or split this function into several and build the operation at the caller. E.g.:
 ```
 github_issues = github.get_issues(page)
 jira_issues = [create_jira_issue(issue) for issue in  github_issues]
 ```

EdDev commented on create_ticket.py:
```diff
@@ -0,0 +1,91 @@
+
+    for page in range(1, GITHUB_MAX_PAGES):
+        issues = github.get_issues(page)
+        if len(issues) == 0:
+            break
```
This is not needed.

oshoval replied:
its needed to stop processing more pages when the pages that are fetched are empty.

EdDev replied:
Hmm.. right, the multi for loops here confused me.

So why this is not part of `github.get_issues()` ? Why do we need to know at this level about "pages"? That is an implementation detail.

EdDev commented on create_ticket.py:
```diff
@@ -0,0 +1,91 @@
+            res = process_issue(jira, github, issue, dryrun)
+            if res:
+                issues_created += 1
+                if issues_created == flood_protection_limit:
+                    print("Flood protection limit reached, exiting")
+                    sys.exit(0)
```
The "flood" should be checked based on the number of github issues read and filtered.
E.g: `issues = github.get_issues(page)[:flood_protection_limit]`

oshoval replied:
you count only for issues that you create, there might be issues that are not sig/nework , or already created, so you don't count them
they might be on different github pages, so it can't be at get_issues which runs per page

EdDev replied:
Ok, so just do not mix the things in one.
Per what you are saying, we need to see:
- Fetch all issues.
- Filter issues per criteria.
- Open Jira issues per the filtered list.

If I understand correctly, the filtered list needs to be limited in size.

In python you can do all this even without wasting memory by using generators (for the outside, it looks like you process lists, but it actually iterates each time).

oshoval replied:
I will refactor it, but note that we should not fetch all issues, because we might get to the threshold before we fetched them all, this is why we break when the threshold condition reached.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,58 @@
+        self.query_url = f"https://api.github.com/repos/{owner}/{repo}/issues"
+        self.headers = {"Authorization": f"token {token}"}
+
+    def get_issues(self, page):
```
Can you rename it to just `issues`?

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,58 @@
+        data = r.json()
+        return data
+
+    def check_labels(self, issue):
```
I do not understand what this does really. You want to filter by label?

oshoval replied:
we filter for example only issues that are `sig/network`
we don't process others
`github_label = get_envvar("GITHUB_LABEL")` is configurable

EdDev replied:
Ok, so this is just an inner helper (and not public). The operation is "filter".

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,58 @@
+                return True
+        return False
+
+    def check_time(self, issue, max_delta):
```
What is this suppose to do? I do not understand.
Is it the issue aging time?

oshoval replied:
we are reading x weeks back
https://github.com/oshoval/github2jira/pull/1/files#diff-63469d0fb919a44c962153d5288c638b1e9e5184a30d002a58eeb8a269e9878aR13

We don't read all the issues, because we don't want to open tickets for too old issues.
On the other hand an issue might be opened, and after just one week assigned a sig/network label, so its better to do scan the history, but for limited time.

EdDev replied:
Same as the label then, this is something internal. And you need to give it a good name, because the current one says nothing meaningful.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,58 @@
+SECONDS_PER_WEEK = 604800
+
+
+class Github:
```
After reading what this class do, I realized that 2 out of its 3 methods are about a single issue.
I would consider creating an issue class as well.

EdDev commented on jiralib.py:
```diff
@@ -0,0 +1,60 @@
+        issues = self.jira.search_issues(query)
+        return len(issues) != 0
+
+    def create_issue(self, repo, issue_id, title, body):
```
It will be nicer if the input is just a github issue object, and not having repos and such details involved at this level.

EdDev reviewed (changes requested):
I am unclear which tests are written to cover all the code. Are these the `main` functions?
I suggest you create a `tests` folder` and create there test modules using pytest that import the production code.

EdDev commented:
Please note that for deployment, it is recommended to use a setup file: https://docs.python.org/3/distutils/setupscript.html

If this is for internal use only, then you do not need it.
If you want to maintain it for others to use, you need it.

oshoval commented:
> Please note that for deployment, it is recommended to use a setup file: https://docs.python.org/3/distutils/setupscript.html
> 
> If this is for internal use only, then you do not need it. If you want to maintain it for others to use, you need it.

Thanks
For now we start with internal use.

oshoval commented:
Thanks

addressed comments,
didn't add yet tests, most of them are pretty tricky because the github status changes (unless we look always on all the history),
and we can't create issues on Jira in tests (we can just check for example a known issue exists).

new files
`config.py`
`ticket.py`
`main.py`
removed the `common.py`

maybe worth to remove github from Ticket class, and just send repo to it, since its the only thing it uses

oshoval commented:
https://github.com/oshoval/github2jira/compare/25d5236..b717bdf
I think this should include all the pushes (some of them hidden in others, and some are small)

oshoval commented:
Refactor the issue class

oshoval commented:
Added unit tests for Issue class

oshoval commented:
added more unit tests
added --issue feature

EdDev commented on Dockerfile:
```diff
@@ -0,0 +1,10 @@
+
+RUN pip install requests jira
+
+CMD [“echo”, “Hello World”]
```
Is this a must?

EdDev commented on LICENSE:
```diff
@@ -186,7 +186,7 @@
       identification within third-party archives.
 
-   Copyright [yyyy] [name of copyright owner]
+   Copyright [2021] [Red Hat, Inc]
```
nit: No need for the parenthesis.

EdDev commented on config.py:
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python3
```
This line is not needed. Same for all other files.

oshoval replied:
from main.py its needed so i can run it with `./main.py` (without need to configure anything more)
from the rest i can remove
it was because the libs were self runnable units before (when they had the __main__ condition)

EdDev commented on config.py:
```diff
@@ -0,0 +1,35 @@
+class Config:
+    def __init__(self, var_names):
+        self.var_names = var_names
+        self.vars = {}
```
You can do this:
`self._vars = {name: None for name in var_names}`

If you make the attribute "private", then you be able to protect it from someone changing it by creating a getter-like method:
```python
@property
def vars():
    return self._vars
```

EdDev commented on config.py:
```diff
@@ -0,0 +1,35 @@
+        self.var_names = var_names
+        self.vars = {}
+
+    def Load(self):
```
Then populate the values:
`self._vars = {name: os.getenv(value) for name, value in self._vars.items()}`

If you want to really "explode" if something is missing, then use a regular for-loop and raise an exception.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,109 @@
+
+from config import Config
+
+SECONDS_PER_WEEK = 604800
```
Can you show how you reached this number? `7 * ....`

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,109 @@
+MAX_DELTA_WEEKS = 4
+
+
+class GithubConfig(Config):
+    def __init__(self):
+        super().__init__(
+            ["GITHUB_TOKEN", "GITHUB_OWNER", "GITHUB_REPO", "GITHUB_LABEL"]
+        )
```
Using inheritance seems an overkill to me.
How about placing the names in a list, and then just do:
At githublib:
```python
_ENV_VAR_NAMES = ["GITHUB_TOKEN", "GITHUB_OWNER", "GITHUB_REPO", "GITHUB_LABEL"]

def config():
    c = Config(_ENV_VAR_NAMES)
    c.Load()
    return c
```

At the usage:
```python
import githublib

github = githublib.Github(githublib.config())
```

oshoval replied:
I prefer it this way, it encapsulate each class what what it needs.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,109 @@
+                return True
+        return False
+
+    def age_relevant(self, max_delta):
```
- The name is not reflecting to me what it does/returns.
- Can't you just return the timestamp and let the caller to its calculation? Asking for the "issue" to tell me if it is in the delta range is odd.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,109 @@
+
+class Github:
+    def __init__(self, cfg):
+        self.owner = cfg.vars["GITHUB_OWNER"]
```
It looks like `"GITHUB_OWNER"` (and the others) needs to be known in more than one place. We need to aim to know its name only once.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,109 @@
+        self.headers = {"Authorization": f"token {cfg.vars['GITHUB_TOKEN']}"}
+
+    def issues(self):
+        for page in range(1, GITHUB_MAX_PAGES):
```
The generators allows you to separate nicely the iteration over the pages and the issues into different functions.
Something like this:
`return (issue for issues in pages(GITHUB_MAX_PAGES) for issue in filter(issues))`

In the above case, `pages` is a generator method (returning with `yield` or `return` when finished earlier).
The filter, returns only relevant issues and is also a generator.
The `(...)` is a comprehension generator, which is just a shorter way to define one.

EdDev commented on jiralib.py:
```diff
@@ -0,0 +1,66 @@
+
+    def create_issue(self, issue):
+        issue_dict = dict()
+        issue_dict["project"] = dict({"id": self.project_id})
```
Why you wrap a dict with a dict?

EdDev commented on tox.ini:
```diff
@@ -0,0 +1,66 @@
+commands =
+    flake8 \
+        --statistics {posargs} \
+        main.py \
+        config.py \
+        githublib.py \
+        jiralib.py \
+        ticket.py \
+        tests/test_github.py \
+        tests/test_config.py
```
Why do you need to specify specific files and not the whole folders? This is not maintainable.
Same for the other occurrence of this.

oshoval replied:
pylint has problems with it (at least on CI)
there is even an open issue about it 
[1] https://github.com/PyCQA/pylint/issues/352

```
pylint run-test: commands[0] | pylint --errors-only main.py config.py githublib.py jiralib.py ticket.py tests/
************* Module tests
tests/__init__.py:1:0: F0010: error while code parsing: Unable to load file tests/__init__.py:
[Errno 2] No such file or directory: 'tests/__init__.py' (parse-error)
ERROR: InvocationError for command /home/runner/work/github2jira/github2jira/.tox/pylint/bin/pylint --errors-only main.py config.py githublib.py jiralib.py ticket.py tests/ (exited with code 1)
```
maybe just need to add the __init__.py file ?
https://stackoverflow.com/questions/48024049/pylint-raises-error-if-directory-doesnt-contain-init-py-file
(the link [1] is also from this discussion)

EDIT - fixed by adding the __init__.py file

fixed for the 2nd

EdDev reviewed (changes requested):
This is a partial review.

Please also add the test run to the tox and then to the CI.

oshoval commented:
thanks, addressed comments, 
working on adding to CI (might need to merge it before, but maybe not, lets see)

oshoval commented:
CI runs now per PR (will make mandatory for merges)
https://github.com/oshoval/github2jira/blob/main/.github/workflows/main.yml

(tweaks might be needed to the above yml)

oshoval commented:
refactor the config according comment

oshoval commented:
fixed pylint to use `tests/` on CI as well
updated actions to be little cleaner

the next push is in order to support more python versions

EdDev commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,41 @@
+# This is a basic workflow to help you get started with Actions
```
Please remove all the comments from the original template.

EdDev commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,41 @@
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install tox
```
This is not part of the setup-python aready?

oshoval replied:
removing first line, but tox is needed

EdDev commented on config.py:
```diff
@@ -0,0 +1,35 @@
+        for var_name in self._vars.keys():
+            value = os.getenv(var_name)
+            if value is None:
+                raise Exception(f"Error: cant find {var_name}")
```
- `can't`
- Consider using a more precise error, either a custom one or just one from [here](https://docs.python.org/3/library/exceptions.html).

EdDev commented on config.py:
```diff
@@ -0,0 +1,35 @@
+            if value is None:
+                raise Exception(f"Error: cant find {var_name}")
+            self._vars[var_name] = value
+        return self
```
Why do you return it? I see no one uses it and I always find it odd to have this "train" usage option.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,115 @@
+    def __init__(self, issue):
+        self.issue = issue
+
+    def filter(self, expected_label):
```
This one looks more like `has_label` to me.
It does not seem to be under the right object, because why should a single issue contain this extra functionality at all.

It makes sense to me that an issue will expose its label (like the other property methods do), and for some other method from another object to perform this filtering (e.g. a issues pool).

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,115 @@
+
+class Github:
+    def __init__(self, cfg):
+        owner = cfg.vars["GITHUB_OWNER"]
+        repo = cfg.vars["GITHUB_REPO"]
```
This is repeating the env var names again, which is not nice. I suggest you create properties to the `Config` for each variable, this way the name of the variable is preserved in one place only, others will just call the property methods on the config.

oshoval replied:
i use those specific two only in the __init__ itself, it is just so the line won't be long
there is no need to create vars for them

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,115 @@
+            if len(issues) == 0:
+                return
+
+            yield issues
```
I think it will be simpler if you do:
```
for issue in issues:
    yield issue
```

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,115 @@
+
+            yield issues
+
+    def filter(self, issues):
```
I would generalize it on one hand and specialize it on the other.
It can be a regular function that receives the issues and the label to filter by. And it is then called `filter_by_label`.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,115 @@
+            if issue.filter(self.expected_label) and filter_by_time(
+                issue, MAX_DELTA_WEEKS
+            ):
+                yield (issue)
```
What is the parenthesis for?

oshoval replied:
black does it the make the line shorter

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,115 @@
+            ):
+                yield (issue)
+
+    def get_issue(self, issue_id):
```
I do not understand this one now. We have `issues`, `filter` and now this one that returns issues, I am confused.

EdDev commented on githublib.py:
```diff
@@ -0,0 +1,115 @@
+        return Issue(issue)
+
+
+def filter_by_time(issue, max_delta_weeks):
```
"filter" is usually associated to a collection, but you seem to check a specific condition for a specific item.
The name should reflect that you return a boolean in such a case.

You could make it a collection, by passing the collection and returning the filtered one back.

EdDev commented on jiralib.py:
```diff
@@ -0,0 +1,65 @@
+            basic_auth=(cfg.vars["JIRA_USERNAME"], cfg.vars["JIRA_TOKEN"]),
+        )
+
+    def issue_exists(self, issue):
```
It took me some time to figure out the argument is about a github issue. Can you rename it to reflect that?

EdDev commented on jiralib.py:
```diff
@@ -0,0 +1,65 @@
+        return len(issues) != 0
+
+    def create_issue(self, issue):
+        issue_dict = dict()
```
Please do not embed the type name in the variable name. You can call it `issue_data` for example.

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,44 @@
+import mock
+import pytest
+
+sys.path.insert(1, sys.path[0] + "/..")
```
What is this?

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,44 @@
+import sys
+import os
+
+import mock
```
You better use the one provided by the stdlib.

EdDev commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+requests
+jira
+mock
```
I do not think you need this one.

oshoval replied:
had to add it as some tests used it,
if it won't be needed at the end i will remove,
without it the `import mock` failed

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,44 @@
+import githublib  # noqa: E402
+
+
+def mockenv(**envvars):
+    return mock.patch.dict(os.environ, envvars)
```
This is very odd. I think you can decorate using `mock.patch` directly, I do not understand why this wrapper is needed.

oshoval replied:
it is a wrapper which i find cleaner
https://newbedev.com/python-mock-patch-os-environ-and-return-value

EdDev replied:
It is not a standard way to do things and it is very confusing to a new reader.
But I do not want to argue about this, I leave it up to you.

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,44 @@
+    GITHUB_REPO="repo",
+    GITHUB_LABEL="sig/network",
+)
+def test_github():
```
Please perform specific tests and name them per what they do.

EdDev commented on ticket.py:
```diff
@@ -0,0 +1,44 @@
+FLOOD_PROTECTION_LIMIT = 3
+
+
+class Ticket:
```
I do not understand what this represents.

oshoval replied:
encapsulates the higher level of ticker creation:
ticket create and flood protection,
can be flatten if needed

oshoval commented:
Thanks, addressed some as we spoke (resolved ones), working on the rest

oshoval commented:
addressed rest of the comments

EdDev commented on github2jira/jiralib.py:
```diff
@@ -0,0 +1,104 @@
+        issue_data["summary"] = f"[GITHUB:{repo}-{id}] {title}"
+        issue_data["description"] = git_issue.url
+        issue_data["issuetype"] = {"name": "Task"}
+        if self.component != "":
+            issue_data["components"] = [{"name": self.component}]
```
Are you sure the key should not be reported at all? Usually a placeholder is given (e.g. an empty list).

oshoval replied:
Yep, moreover it depends on the Jira you are using
tested it versus 2 Jira domains, one doesn't even support components on this query.
On both there are parameters which are optional and can be left unassigned.
The mandatory parameters will fail the query in case not supplied.

EdDev commented on github2jira/jiralib.py:
```diff
@@ -0,0 +1,104 @@
+        print(f"Created issue {issue_url} for {git_issue.url}")
+
+    def _create_issue_data(self, git_issue):
+        repo = git_issue.repo
+        id = git_issue.id
+        title = git_issue.title
+
+        issue_data = dict()
+        issue_data["project"] = {"id": self.project_id}
+        issue_data["summary"] = f"[GITHUB:{repo}-{id}] {title}"
+        issue_data["description"] = git_issue.url
+        issue_data["issuetype"] = {"name": "Task"}
```
You can try this:
```
issue_data = {
    "project": {"id": self.project_id}
    "summary": ...
}
```

It is usually better.

oshoval commented:
addressed comment

oshoval commented:
added some more tests, and tests/common lib

EdDev commented on github2jira/githublib.py:
```diff
@@ -0,0 +1,123 @@
+                yield Issue(issue)
+
+
+def issue_in_window(issue, max_delta_weeks):
```
nit: `issue_in_time_window`

EdDev commented on github2jira/ticketmanager.py:
```diff
@@ -0,0 +1,42 @@
+        else:
+            print("Issue for", issue.url, "already exists")
+
+        return False
```
Returning a boolean on a "create" operation seems too odd.
If there is a problem, it should raise an exception.

Why do you need this?

oshoval replied:
Only for the unit test so it will count how many it "created" and how many it didnt
not for indication of an error

EdDev replied:
Hmm.. tests should make it look better, not in such an odd behavior.
I would expect `create` to either return nothing or the issue itself. In this case, returning nothing makes more sense, as no one in the production code needs it.

For the tests, you do have a value to check against: The counter.

EdDev commented on test_ticketmanager.py:
```diff
@@ -0,0 +1,47 @@
+from github2jira.githublib import Issue
+
+
+class MockJira:
```
nit: `JiraMock`.

EdDev commented on tests/common.py:
```diff
@@ -0,0 +1,39 @@
+# This file is part of the github2jira project
```
`common` is not a valid name for a module. Please try and see if you can name it with something that hints about the content.

oshoval replied:
`testdata.py` ?

EdDev commented on tests/common.py:
```diff
@@ -0,0 +1,39 @@
+COMPONENT = "dummy"
+
+
+def get_raw_issue():
```
Please drop the `get`, I do not see how it fits here.

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,42 @@
+
+import os
+
+from unittest import mock
```
This is an stdlib, so it belong to the group above (and not with the 3rd party `pytest` one).

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,42 @@
+    GITHUB_LABEL="sig/network",
+)
+def test_config_load():
+    githublib.config()
```
I do not understand what you are testing here. You have not checked if it worked (the results).

oshoval replied:
just that it doesnt throw an exception, this is one of the methods unless we want to be more explicit (on it)

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,42 @@
+    githublib.config()
+
+
+def test_negative_config_load():
```
negative? Please express what you are testing explicitly. By reading this test, I just see you expect it to raise, but I do not know why.

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,42 @@
+
+def test_negative_config_load():
+    with pytest.raises(Exception):
+        githublib.config()
```
I think you are raising `NameError` and not a general exception, please be specific as there may be other errors that can raise as well (and we do not want to catch them here).

EdDev commented on tests/test_github.py:
```diff
@@ -0,0 +1,101 @@
+
+import os
+
+from unittest import mock
```
same, this is a stdlib.

EdDev commented on tests/test_jiralib.py:
```diff
@@ -0,0 +1,65 @@
+
+import os
+
+from unittest import mock
+
+from unittest.mock import patch
```
stdlib

It is also preferable not to loose the `mock` naming, as it gives context to what `patch` refers to.

EdDev commented on tests/test_jiralib.py:
```diff
@@ -0,0 +1,65 @@
+        return
+
+
+@pytest.fixture
+def mocked_jira():
+    return MockedJIRA()
```
I do not think you need this one.. it gives you nothing.

oshoval replied:
we need to mock jira.JIRA so it won't try to connect the real jira in the constructor of jirablib.Jira

ref
https://github.com/oshoval/github2jira/pull/1#discussion_r758122518

EdDev commented on tests/test_jiralib.py:
```diff
@@ -0,0 +1,65 @@
+)
+@patch("jira.JIRA")
+def test_initialization(self, mocked_jira):
+    from github2jira.jiralib import Jira, config
```
It is a bad practice to import inside functions, please avoid it.

oshoval replied:
yep dirty leftover (interesting it affects the patching if i move it)

EdDev commented on tests/test_jiralib.py:
```diff
@@ -0,0 +1,65 @@
+    JIRA_PROJECT_ID=common.PROJECT_ID,
+    JIRA_COMPONENT=common.COMPONENT,
+)
+@patch("jira.JIRA")
```
I do not understand what you are patching here. Usually patching is done with an alternative object (e.g. `MockedJIRA`).

oshoval replied:
monkey patching the jira.JIRA constructor, each instance that it will need will be replaced with MockedJIRA
so this call wont try to reach server (because it does otherwise)
https://github.com/oshoval/github2jira/pull/1/files#diff-64115fdc7e5e237a82ce6b47df31a57b5f51c7300f94cee8be1d3aab18c52ae4R58

anyhow updating

EdDev commented on tests/test_github.py:
```diff
@@ -0,0 +1,101 @@
+    raw_issue = get_raw_issue()
+    issue = githublib.Issue(raw_issue)
+
+    assert issue.repo == "repo"
```
Please ref the expected result from the place it is defined, do not repeat it yourself here as it is unclear why this is the value and not something else.

In some places you did it right.

EdDev commented on tests/test_jiralib.py:
```diff
@@ -0,0 +1,65 @@
+    JIRA_COMPONENT=common.COMPONENT,
+)
+@patch("jira.JIRA")
+def test_initialization(self, mocked_jira):
```
I do not understand what does it mean to test initialization.

EdDev reviewed (changes requested):
The production code seems good, I only had small comments about it, nothing major.

The unit tests require more work.
Although I will not insist on it, please note that the unit tests are still suppose to check scenarios and not individual methods.  I mean: 
- checking that you can filter the relevant github issues.
- Opening a Jira ticket, based on some data.
- Failing to open a Jira ticket or contacting github.
- Not finding any relevant issues on github to process.

Etc...

oshoval commented:
> The production code seems good, I only had small comments about it, nothing major.
> 
> The unit tests require more work. Although I will not insist on it, please note that the unit tests are still suppose to check scenarios and not individual methods. I mean:
> 
> * checking that you can filter the relevant github issues.
> * Opening a Jira ticket, based on some data.
> * Failing to open a Jira ticket or contacting github.
> * Not finding any relevant issues on github to process.
> 
> Etc...

Thanks, 
some of those sound as e2e,
I prefer to not develop this project any deeper right now in aspect of tests, and provide as is.
In case we decide we can improve it later.

oshoval commented:
addressed most of the comments

oshoval commented:
Thanks, addressed rest of the comments (unless unresolved)

oshoval commented:
since some Jira servers will deprecate basic auth and move to PAT, support it

EdDev commented on github2jira/githublib.py:
```diff
@@ -0,0 +1,123 @@
+#
+
+import time
+import requests
```
this is not a stdlib.

(I think datetime is, right?)

EdDev commented on main.py:
```diff
@@ -0,0 +1,62 @@
+    ticket_manager = TicketManager(jira)
+
+    if args.issue is None:
+        process_issues(github, ticket_manager)
```
`process_issue` is not a really descriptive, how about: `create_jira_tickets`?
The other one is a simple, `create_jira_ticket`.

EdDev commented on test_ticketmanager.py:
```diff
@@ -0,0 +1,47 @@
+    issues_created = 0
+    for i in range(FLOOD_PROTECTION_LIMIT * 3):
+        if ticket_manager.create(git_issue):
+            issues_created += 1
```
You already have a counter for this: `ticket_manager.issues_created`.
I guess this comment comes hand in hand with the previous one about returning a boolean from the `create` method.

EdDev commented on tests/test_config.py:
```diff
@@ -0,0 +1,45 @@
+    GITHUB_LABEL="sig/network",
+)
+def test_config_load():
+    try:
+        githublib.config()
+    except NameError:
+        pytest.fail("Unexpected NameError, check env var names")
```
My previous comment was not intended for you to do this, the test would have failed if an exception raises, so this is not needed.

I was referring to the check of the content you expect to have in the returned object. It makes sense to check it.

EdDev commented on tests/test_github.py:
```diff
@@ -0,0 +1,101 @@
+
+    issues = list(github.issues())
+
+    assert len(issues) == 3
+    assert issues[0].url == "https://github.com/owner/repo/issues/10"
+    assert issues[1].url == "https://github.com/kubevirt/kubevirt/issues/11"
+    assert issues[2].url == "https://github.com/kubevirt/kubevirt/issues/13"
```
It is preferred to assert once. In case of failure, we are interested to see the big picture and quickly understand what happened.

Please assert the list directly, expecting to see exactly 3 entries with the relevant content.

EdDev commented on tests/test_github.py:
```diff
@@ -0,0 +1,101 @@
+        status=200,
+    )
+
+    assert github.issue_by_id(issue_id) is not None
+    assert github.issue_by_id(issue_id_na) is None
```
Seems to be two different tests to me.
You can use [pytest parameterize](https://docs.pytest.org/en/6.2.x/parametrize.html) if it fits or just two tests.

EdDev reviewed (changes requested):
> some of those sound as e2e,

I do not think they are, as the intention is to unit test scenarios that have meaning (if possible).

> I prefer to not develop this project any deeper right now in aspect of tests, and provide as is.
> In case we decide we can improve it later.

Not investing in tests is raising the risk you still have bugs or worse, that you may introduce bugs in the future.
As you are the one who needs to maintain this, it is your burden to carry so I will not push you to do anything beyond what you want.

oshoval commented:
> I do not think they are, as the intention is to unit test scenarios that have meaning (if possible).

Some aren't possible.

For example `jira.create_issue`, i already checked the jira create data but
beside this, the only thing the function does is call the real Jira lib, which we can't do in unit test.
the real jira function even doesn't return a value (just assert in case of error, and i don't try to catch it anyhow even on main)

Anyhow as we said, I prefer to first publish an initial version, and upon needs to develop it,
maintaining of it is very easy, and i don't see we need to invest it any deeper right now.

oshoval commented:
thanks, addressed comments

EdDev reviewed (approved):
Nice work!

oshoval commented:
Thanks for the great review Edy
